src/menus/PluginMenus.cpp

```diff
@@ -28,10 +28,381 @@
 
+
+using EffectsMenuGroups = std::vector<std::pair<TranslatableString, std::vector<TranslatableString>>>;
+
+struct EffectsMenuGroupsHandler : XMLTagHandler
+{
+   struct EffectsHandler : public XMLTagHandler
+   {
+      std::optional<std::string> textContent;
+      std::vector<TranslatableString>& effects;
+
+      EffectsHandler(std::vector<TranslatableString>& effects) : effects(effects) { }
+
+      bool HandleXMLTag(const std::string_view& tag, const AttributesList& attrs) override { return true; }
+      void HandleXMLContent(const std::string_view& text) override { textContent = text; }
+      void HandleXMLEndTag(const std::string_view& tag) override
+      {
+         if(textContent.has_value() && tag == "Effect")
+         {
+            effects.emplace_back(TranslatableString { *textContent, {} });
+            textContent = {};
+         }
+      }
+      XMLTagHandler* HandleXMLChild(const std::string_view& tag) override
+      {
+         if(tag == "Effect")
+            return this;
+         return nullptr;
+      }
+
+   };
+
+   struct GroupHandler : public XMLTagHandler
+   {
+      std::optional<std::string> textContent;
+      std::unique_ptr<EffectsHandler> effectsHandler;
+      std::pair<TranslatableString, std::vector<TranslatableString>>& group;
+
+      GroupHandler(std::pair<TranslatableString, std::vector<TranslatableString>>& group) : group(group) { }
+
+      bool HandleXMLTag(const std::string_view& tag, const AttributesList& attrs) override { return true; }
+      void HandleXMLContent(const std::string_view& text) override { textContent = text; }
+      void HandleXMLEndTag(const std::string_view& tag) override
+      {
+         if(tag == "Name")
+         {
+            group.first = TranslatableString { *textContent, { } };
+            textContent = { };
+         }
+      }
+      XMLTagHandler* HandleXMLChild(const std::string_view& tag) override
+      {
+         if(tag == "Effects")
+         {
+            effectsHandler = std::make_unique<EffectsHandler>(group.second);
+            return &*effectsHandler;
+         }
+         if(tag == "Name")
+            return this;
+
+         return nullptr;
+      }
+   };
+
+   EffectsMenuGroups& groups;
+   std::unique_ptr<GroupHandler> groupHandler;
+
+   EffectsMenuGroupsHandler(EffectsMenuGroups& groups) : groups(groups) { }
+
+   bool HandleXMLTag(const std::string_view& tag, const AttributesList& attrs) override { return true; }
+
+   XMLTagHandler* HandleXMLChild(const std::string_view& tag) override
+   {
+      if(tag == "Group")
+      {
+         groups.resize(groups.size() + 1);
+         groupHandler = std::make_unique<GroupHandler>(groups.back());
+         return &*groupHandler;
+      }
+      return nullptr;
+   }
+};
+
+EffectsMenuGroups LoadEffectsMenuGroups(const wxString& path)
+{
+   EffectsMenuGroups result;
+   EffectsMenuGroupsHandler handler(result);
+
+   XMLFileReader reader;
+   reader.Parse(&handler, path);
+   return result;
```

Comment:
This function nicely encapsulates the decisions about default grouping and sorting of effects.  To define some different policy in a later version should change only what happens in here.

Maybe a static variable can hold the results of the parse so that the .xml file need not be read more than once each session.

But didn't we agree to sort effect names in each group by their translations?  Easily added here.

Did we also want to sort the groups by their translations?

Re-sorting of names in the statically duration results might be repeated in case the user changes languages and the menus must be rebuilt.

Reply:
Maybe we only want to sort each list by translation, but keep the names ordered as in the resource file.  Whatever the decision, it is easy to change it just here.

Reply:
For me on Mac, the sorting doesn't work.

The file is at
/Users/paullicameli/Github/audacity2/build/bin/Debug/Audacity.app/Contents/res/effects_menu_defaults.xml

The program looks for a file at
/Users/paullicameli/Github/audacity2/build/bin/Debug/Audacity.app/Contents/MacOS/res/effects_menu_defaults.xml

Reply:
I moved the .xml to the expected place in the tree and was then able to test the changes and verify that switching grouping criteria in Effects preferences does as expected.